File: Python/shared/baseline_item.py
# Exact full port of BaselineItem.cs (abstract base)

import xml.etree.ElementTree as ET

class BaselineItem:
    def __init__(self):
        self.m_sSourceFile: str = ""
        self.m_sGpoName: str = ""
        self.m_sEvalOperator: str = ""
        self.m_sEvalValues: str = ""
        self.m_sBaselineValue: str = ""
        self.m_bValueFound: bool = True
        self.m_bEvalPassed: bool = False
        self.m_bInvalidComparison: bool = False

    def SortKey(self) -> str:
        raise NotImplementedError("Subclasses must implement SortKey")

    def ReadStdElems(self, node: ET.Element):
        self.m_sSourceFile = self._optional_inner_text(node, "SourceFile")
        self.m_sGpoName = self._optional_inner_text(node, "PolicyName")
        self.m_sEvalOperator = self._optional_inner_text(node, "EvalOperator")
        if not self.m_sEvalOperator:
            self.m_sEvalOperator = "Equal"
        self.m_sEvalValues = self._optional_inner_text(node, "EvalValues")

        flag = self._optional_boolean(node, "ValueNotFound")
        self.m_bValueFound = not flag if flag is not None else True

        flag = self._optional_boolean(node, "EvalResult")
        if flag is not None:
            self.m_bEvalPassed = flag

        flag = self._optional_boolean(node, "InvalidComparison")
        if flag is not None:
            self.m_bInvalidComparison = flag

    def _optional_boolean(self, node: ET.Element, child_name: str):
        elem = node.find(child_name)
        if elem is not None and elem.text:
            return elem.text.strip().lower() == "true"
        return None

    def _optional_inner_text(self, node: ET.Element, child_name: str) -> str:
        elem = node.find(child_name)
        return elem.text.strip() if elem is not None and elem.text else ""

    def CreateResultsElements(self, x_output_doc: ET.Element, e_base_element: ET.Element):
        # EvalResult
        elem = ET.SubElement(e_base_element, "EvalResult")
        elem.text = str(self.m_bEvalPassed).lower()
        # BaselineValue
        elem = ET.SubElement(e_base_element, "BaselineValue")
        elem.text = self.m_sBaselineValue
        # EvalOperator
        elem = ET.SubElement(e_base_element, "EvalOperator")
        elem.text = self.m_sEvalOperator
        if self.m_sEvalValues:
            elem = ET.SubElement(e_base_element, "EvalValues")
            elem.text = self.m_sEvalValues
        # ValueNotFound
        elem = ET.SubElement(e_base_element, "ValueNotFound")
        elem.text = str(not self.m_bValueFound).lower()
        # InvalidComparison
        elem = ET.SubElement(e_base_element, "InvalidComparison")
        elem.text = str(self.m_bInvalidComparison).lower()

    def Evaluate(self, sources, x_output_doc: ET.Element):
        raise NotImplementedError("Subclasses must implement Evaluate")

    def IgnoreAlwaysPass(self) -> bool:
        return self.m_sEvalOperator == "Ignore-AlwaysPass"

    def ImplementIgnoreAlwaysPass(self) -> bool:
        if self.IgnoreAlwaysPass():
            self.m_bEvalPassed = True
            self.m_bInvalidComparison = False
            return True
        return False

File: CS/PolicyAnalyzer/SharedCode/BaselineItem.cs
using System;
using System.Xml;

namespace SharedCode
{
	// Token: 0x0200003A RID: 58
	internal abstract class BaselineItem
	{
		// Token: 0x060000A7 RID: 167
		public abstract string SortKey();

		// Token: 0x060000A8 RID: 168 RVA: 0x00008FF4 File Offset: 0x000071F4
		internal void ReadStdElems(XmlNode node)
		{
			this.m_sSourceFile = this.OptionalInnerText(node, "SourceFile");
			this.m_sGpoName = this.OptionalInnerText(node, "PolicyName");
			this.m_sEvalOperator = this.OptionalInnerText(node, "EvalOperator");
			if (this.m_sEvalOperator.Length == 0)
			{
				this.m_sEvalOperator = "Equal";
			}
			this.m_sEvalValues = this.OptionalInnerText(node, "EvalValues");
			bool flag;
			if (this.OptionalBoolean(node, "ValueNotFound", out flag))
			{
				this.m_bValueFound = !flag;
			}
			else
			{
				this.m_bValueFound = true;
			}
			if (this.OptionalBoolean(node, "EvalResult", out flag))
			{
				this.m_bEvalPassed = flag;
			}
			if (this.OptionalBoolean(node, "InvalidComparison", out flag))
			{
				this.m_bInvalidComparison = flag;
			}
		}

		// Token: 0x060000A9 RID: 169 RVA: 0x000090B4 File Offset: 0x000072B4
		private bool OptionalBoolean(XmlNode node, string childName, out bool bValue)
		{
			bValue = false;
			XmlElement xmlElement = node[childName];
			return xmlElement != null && bool.TryParse(xmlElement.InnerText, out bValue);
		}

		// Token: 0x060000AA RID: 170 RVA: 0x000090E0 File Offset: 0x000072E0
		internal string OptionalInnerText(XmlNode node, string childName)
		{
			XmlElement xmlElement = node[childName];
			if (xmlElement != null)
			{
				return xmlElement.InnerText;
			}
			return string.Empty;
		}

		// Token: 0x060000AB RID: 171 RVA: 0x00009104 File Offset: 0x00007304
		internal void CreateResultsElements(XmlDocument xOutputDoc, XmlElement eBaseElement)
		{
			XmlElement xmlElement = xOutputDoc.CreateElement("EvalResult");
			xmlElement.InnerText = this.m_bEvalPassed.ToString();
			eBaseElement.AppendChild(xmlElement);
			XmlElement xmlElement2 = xOutputDoc.CreateElement("BaselineValue");
			xmlElement2.InnerText = this.m_sBaselineValue;
			eBaseElement.AppendChild(xmlElement2);
			XmlElement xmlElement3 = xOutputDoc.CreateElement("EvalOperator");
			xmlElement3.InnerText = this.m_sEvalOperator;
			eBaseElement.AppendChild(xmlElement3);
			if (this.m_sEvalValues.Length > 0)
			{
				XmlElement xmlElement4 = xOutputDoc.CreateElement("EvalValues");
				xmlElement4.InnerText = this.m_sEvalValues;
				eBaseElement.AppendChild(xmlElement4);
			}
			XmlElement xmlElement5 = xOutputDoc.CreateElement("ValueNotFound");
			xmlElement5.InnerText = (!this.m_bValueFound).ToString();
			eBaseElement.AppendChild(xmlElement5);
			XmlElement xmlElement6 = xOutputDoc.CreateElement("InvalidComparison");
			xmlElement6.InnerText = this.m_bInvalidComparison.ToString();
			eBaseElement.AppendChild(xmlElement6);
		}

		// Token: 0x060000AC RID: 172
		public abstract void Evaluate(EffectiveStateSources sources, XmlDocument xOutputDoc);

		// Token: 0x060000AD RID: 173 RVA: 0x00002C69 File Offset: 0x00000E69
		internal bool IgnoreAlwaysPass()
		{
			return this.m_sEvalOperator == "Ignore-AlwaysPass";
		}

		// Token: 0x060000AE RID: 174 RVA: 0x00002C7B File Offset: 0x00000E7B
		internal bool ImplementIgnoreAlwaysPass()
		{
			if (this.IgnoreAlwaysPass())
			{
				this.m_bEvalPassed = true;
				this.m_bInvalidComparison = false;
				return true;
			}
			return false;
		}

		// Token: 0x04000668 RID: 1640
		internal string m_sSourceFile;

		// Token: 0x04000669 RID: 1641
		internal string m_sGpoName;

		// Token: 0x0400066A RID: 1642
		internal string m_sEvalOperator;

		// Token: 0x0400066B RID: 1643
		internal string m_sEvalValues;

		// Token: 0x0400066C RID: 1644
		internal string m_sBaselineValue = string.Empty;

		// Token: 0x0400066D RID: 1645
		internal bool m_bValueFound;

		// Token: 0x0400066E RID: 1646
		internal bool m_bEvalPassed;

		// Token: 0x0400066F RID: 1647
		internal bool m_bInvalidComparison;
	}
}
